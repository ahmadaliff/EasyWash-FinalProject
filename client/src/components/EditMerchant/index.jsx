import Proptypes from 'prop-types';
import { Dialog, DialogActions, DialogContent, DialogTitle, FormControl, IconButton } from '@mui/material';
import { Close } from '@mui/icons-material';
import { FormattedMessage, injectIntl } from 'react-intl';

import classes from '@components/EditMerchant/style.module.scss';
import { useState } from 'react';
import { useForm } from 'react-hook-form';
import { useDispatch } from 'react-redux';
import InputRHF from '@components/InputRHF';

import MapLeaflet from '@components/MapLeaflet';
import { actionEditMerchant } from '@pages/MyMerchant/actions';

const EditMerchant = ({ merchant, open, handleClose, intl: { formatMessage } }) => {
  const dispatch = useDispatch();
  const [location, setLocation] = useState();
  const {
    handleSubmit,
    register,
    formState: { errors },
    watch,
  } = useForm();

  const name = watch('name');
  const description = watch('description');

  const onSubmit = (data) => {
    data.location = location ? JSON.stringify(location) : JSON.parse(merchant?.location);
    dispatch(actionEditMerchant(data));
    handleClose();
  };

  const handleLocation = (loc) => {
    setLocation(loc);
  };

  return (
    <Dialog open={open} onClose={handleClose}>
      <DialogTitle fontSize="larger">
        <FormattedMessage id="app_edit_merchant" />
      </DialogTitle>
      <IconButton
        aria-label="close"
        onClick={handleClose}
        sx={{
          position: 'absolute',
          right: 8,
          top: 8,
        }}
      >
        <Close />
      </IconButton>
      <DialogContent>
        <form onSubmit={handleSubmit(onSubmit)}>
          <h5 className={classes.merchantFormHeader}>Merchant Info</h5>
          <InputRHF
            input={{
              name: 'name',
              required: formatMessage({ id: 'app_merchant_require' }),
              type: 'text',
              label: formatMessage({ id: 'app_merchant_name' }),
              value: merchant?.name,
            }}
            register={register}
            errors={errors}
          />
          <FormControl fullWidth>
            <div className={classes.inputLabel}>
              <FormattedMessage id="app_location" />
              <MapLeaflet handleLocation={handleLocation} islocated={merchant && JSON.parse(merchant?.location)} />
            </div>
          </FormControl>
          <FormControl className={classes.inputContainer}>
            <label htmlFor="description" className={classes.inputLabel}>
              <FormattedMessage id="app_description" />
              <textarea
                className={`${classes.inputTextArea} ${errors.description && classes.inputError}`}
                {...register('description', {})}
                defaultValue={merchant?.description}
              />
            </label>
          </FormControl>
          <DialogActions>
            <button type="submit" className={classes.buttonSubmit} disabled={!name || !description}>
              <FormattedMessage id="app_edit_merchant" />
            </button>
          </DialogActions>
        </form>
      </DialogContent>
    </Dialog>
  );
};
EditMerchant.propTypes = {
  merchant: Proptypes.object,
  intl: Proptypes.object,
  open: Proptypes.bool,
  handleClose: Proptypes.func,
};

export default injectIntl(EditMerchant);
