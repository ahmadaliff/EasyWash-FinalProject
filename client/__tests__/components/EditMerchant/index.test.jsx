import EditMerchant from '@components/EditMerchant';
import { act, fireEvent, render, screen } from '@utils/testHelper';

let wrapper;
let isOpen = true;
let inputName;
let inputDescription;

jest.mock('@components/MapLeaflet', () => () => <div>mapleafletMock</div>);

jest.mock('react-redux', () => ({
  ...jest.requireActual('react-redux'),
  useDipatch: () => jest.fn(),
}));

const mockProps = {
  open: isOpen,
  intl: { formatMessage: jest.fn() },
  handleClose: () => (isOpen = false),
  merchant: {
    name: 'merchant',
    location: '{"lat": -6.224934545453234, "lng": 106.86266439802276 }',
    description: 'description',
  },
};

beforeEach(() => {
  act(() => {
    wrapper = render(<EditMerchant {...mockProps} />);
  });
});

describe('Edit Merchant Component', () => {
  test('Correct render', () => {
    const { getByTestId } = wrapper;
    expect(getByTestId('dialog-edit-merchant')).toBeInTheDocument();
    expect(getByTestId('button-close')).toBeInTheDocument();
  });
  test('Should be close the dialog', () => {
    const { getByTestId } = wrapper;
    const button = getByTestId('button-close');
    fireEvent.click(button);
    expect(isOpen).toBe(false);
  });
  test('Should be submit the dialog', () => {
    const { getByTestId } = wrapper;
    inputName = screen.getByLabelText('Nama Laundry');
    inputDescription = getByTestId('input-description');
    fireEvent.change(inputName, { target: { value: 'Laundry Mantap' } });
    fireEvent.change(inputDescription, { target: { value: 'ini adlaah deskripsi' } });
    expect(inputName.value).toBe('Laundry Mantap');
    expect(inputDescription.value).toBe('ini adlaah deskripsi');
    const button = getByTestId('button-submit');
    act(() => {
      fireEvent.click(button);
    }).then(() => {
      expect(isOpen).toBe(false);
    });
  });
  test('Should match with snapshot', () => {
    expect(wrapper).toMatchSnapshot();
  });
});
