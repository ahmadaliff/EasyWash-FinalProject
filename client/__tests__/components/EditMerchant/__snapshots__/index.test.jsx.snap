// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Edit Merchant Component Should match with snapshot 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body
    style="padding-right: 1024px; overflow: hidden;"
  >
    <div
      aria-hidden="true"
    />
    <div
      class="MuiDialog-root MuiModal-root css-zw3mfo-MuiModal-root-MuiDialog-root"
      data-testid="dialog-edit-merchant"
      role="presentation"
    >
      <div
        aria-hidden="true"
        class="MuiBackdrop-root MuiModal-backdrop css-yiavyu-MuiBackdrop-root-MuiDialog-backdrop"
        style="opacity: 1; webkit-transition: opacity 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms; transition: opacity 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;"
      />
      <div
        data-testid="sentinelStart"
        tabindex="0"
      />
      <div
        class="MuiDialog-container MuiDialog-scrollPaper css-hz1bth-MuiDialog-container"
        role="presentation"
        style="opacity: 1; webkit-transition: opacity 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms; transition: opacity 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;"
        tabindex="-1"
      >
        <div
          aria-labelledby=":r3:"
          class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation24 MuiDialog-paper MuiDialog-paperScrollPaper MuiDialog-paperWidthSm css-1t1j96h-MuiPaper-root-MuiDialog-paper"
          role="dialog"
        >
          <h2
            class="MuiTypography-root MuiTypography-h6 MuiDialogTitle-root css-10wwfxj-MuiTypography-root-MuiDialogTitle-root"
            id=":r3:"
          >
            Ubah Laundry
          </h2>
          <button
            aria-label="close"
            class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium css-qsfli3-MuiButtonBase-root-MuiIconButton-root"
            data-testid="button-close"
            tabindex="0"
            type="button"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
              data-testid="CloseIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"
              />
            </svg>
            <span
              class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
            />
          </button>
          <div
            class="MuiDialogContent-root css-ypiqx9-MuiDialogContent-root"
          >
            <form>
              <h5
                class="merchantFormHeader"
              >
                Merchant Info
              </h5>
              <div
                class="MuiFormControl-root inputContainer css-1nrlq1o-MuiFormControl-root"
                data-testid="inputRHF"
              >
                <label
                  class="inputLabel"
                  for="name"
                >
                  Nama Laundry
                  <input
                    class="input undefined"
                    id="name"
                    name="name"
                    type="text"
                    value="merchant"
                  />
                </label>
              </div>
              <div
                class="MuiFormControl-root MuiFormControl-fullWidth css-q8hpuo-MuiFormControl-root"
              >
                <div
                  class="inputLabel"
                >
                  Lokasi
                  <div>
                    mapleafletMock
                  </div>
                </div>
              </div>
              <div
                class="MuiFormControl-root inputContainer css-1nrlq1o-MuiFormControl-root"
              >
                <label
                  class="inputLabel"
                  for="description"
                >
                  Deskripsi
                  <textarea
                    class="inputTextArea undefined"
                    data-testid="input-description"
                    name="description"
                  >
                    description
                  </textarea>
                </label>
              </div>
              <div
                class="MuiDialogActions-root MuiDialogActions-spacing css-knqc4i-MuiDialogActions-root"
              >
                <button
                  class="buttonSubmit"
                  data-testid="button-submit"
                  disabled=""
                  type="submit"
                >
                  Ubah Laundry
                </button>
              </div>
            </form>
          </div>
        </div>
      </div>
      <div
        data-testid="sentinelEnd"
        tabindex="0"
      />
    </div>
  </body>,
  "container": <div
    aria-hidden="true"
  />,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
